src/main/java/org/truffleruby/core/regexp/TruffleRegexpNodes.java

```diff
@@ -75,34 +76,75 @@
 
-        if (encodingConversion && regex.getEncoding() != enc) {
-            EncodingCache encodingCache = regexp.cachedEncodings;
-            regex = encodingCache.getOrCreate(enc, e -> makeRegexpForEncoding(context, regexp, e, currentNode));
+        public static CheckEncodingNode create() {
+            return TruffleRegexpNodesFactory.CheckEncodingNodeGen.create();
         }
 
-        return regex.matcher(stringBytes, start, stringBytes.length);
-    }
+        public abstract Encoding executeCheckEncoding(RubyRegexp regexp, Object str);
 
-    @TruffleBoundary
-    public static Encoding checkEncoding(RubyRegexp regexp, Encoding strEnc, CodeRange codeRange) {
-        final Encoding regexEnc = regexp.regex.getEncoding();
+        @Specialization(guards = {
+                "!isSameEncoding(regexp, string)",
+                "isUSASCII(regexp, string)"
+        })
+        protected Encoding checkEncodingAsciiOnly(RubyRegexp regexp, Object string) {
+            return USASCIIEncoding.INSTANCE;
+        }
+
+        @Specialization(guards = {
+                "isSameEncoding(regexp, string)"
+        })
+        protected Encoding checkEncodingSameEncoding(RubyRegexp regexp, Object string) {
+            return regexp.regex.getEncoding();
+        }
+
+        @Specialization(guards = {
+                "!isSameEncoding(regexp, string)",
+                "!isUSASCII(regexp, string)",
+                "isFixedEncoding(regexp, string)",
+        })
+        protected Encoding checkEncodingFixedEncoding(RubyRegexp regexp, Object string) {
+            return regexp.regex.getEncoding();
+        }
+
+        @Specialization(guards = {
+                "!isSameEncoding(regexp, string)",
+                "!isUSASCII(regexp, string)",
+                "!isFixedEncoding(regexp, string)"
+        })
+        protected Encoding fallback(RubyRegexp regexp, Object string) {
+            return stringEncoding(string);
+        }
+
+        protected boolean isUSASCII(RubyRegexp regexp, Object string) {
+            return regexp.regex.getEncoding() == USASCIIEncoding.INSTANCE &&
+                    codeRangeNode.execute(stringLibrary.getRope(string)) == CodeRange.CR_7BIT;
+        }
+
+        protected boolean isFixedEncoding(RubyRegexp regexp, Object string) {
+            return regexp.options.isFixed() && stringLibrary.getRope(string).getEncoding().isAsciiCompatible();
+        }
+
+        protected boolean isSameEncoding(RubyRegexp regexp, Object string) {
+            return regexp.regex.getEncoding() == stringLibrary.getRope(string).getEncoding();
+        }
 
-        if (strEnc == regexEnc) {
-            return regexEnc;
-        } else if (regexEnc == USASCIIEncoding.INSTANCE && codeRange == CodeRange.CR_7BIT) {
-            return regexEnc;
-        } else if (strEnc.isAsciiCompatible() && regexp.options.isFixed()) {
-            return regexEnc;
+        protected Encoding stringEncoding(Object string) {
```

Comment:
Seems this should be used in more places in this node, or inlined in its only caller.
I'd use it in more places as it seems nicer than `stringLibrary.getRope(string).getEncoding()` in the other helpers, I'll do that.

Reply:
Thanks. That particular line went through a few different designs (`@CreateCast` with a new `ToRopeNod` amongst others). It looks like I just forgot to clean that one up.

Reply:
One idea here would be to add an extra `Encoding stringEncoding` argument to the execute() method, and add a new method called by other nodes, that would do the stringEncoding() logic. The advantage would be that the compiled code contains only once the logic for stringEncoding(), and it'd be similar conceptually to `@CreateCast`. I'll do that.